### PR TITLE
WIP: Add Elasticsearch support

### PR DESCRIPTION
**Pull request by dkarlovi**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/rollerworks/search/issues/179
| License       | MIT
| Doc PR        | (pending)

https://github.com/rollerworks/search/issues/179

### CODE REVIEW

File: src/Elasticsearch/Extension/SearchExtension.php [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            $this->elasticsearchFactory->createConditionGenerator($condition)
+        );
+
+        foreach ($configuration['mappings'] as $fieldName => $mapping) {
+            $conditionGenerator->registerField($fieldName, $mapping);
+        }
+
+        // TODO: temporary, how to do this better?
+        $identifier = 'id';
+        $search = new Search($this->elasticaClient);
```
Comment (dkarlovi) [inline src/Elasticsearch/Extension/SearchExtension.php]: @sstok this is where the question of responsibility comes in, if you have any suggestions, I'd love to hear them.

Comment (dkarlovi) [reply src/Elasticsearch/Extension/SearchExtension.php]: This should probably be taken from entity metadata. Don't know about composite keys.

Comment (dkarlovi) [reply src/Elasticsearch/Extension/SearchExtension.php]: Problem with this approach is you embed the search functionality into a API Platform extension.

What would be better is to have it be available stand-alone, probably with what FOS Elastica is doing, repository based? Ie. repository implementing an interface with

```php
public function search(SearchCondition $searchCondition): array;
```

with a trait converting the condition using Elasticsearch query generator and returning proper results (as Entities).

**Comment by dkarlovi** [comment]
Related to https://github.com/api-platform/core/issues/991

File: src/Elasticsearch/Extension/SearchExtension.php (hunk 1 above)
Comment (sstok) [reply src/Elasticsearch/Extension/SearchExtension.php]: Hmm. I agree this functionality is better suited in a separate class/package, but it's properly easiest (for now) to keep it in this package and see what works and can be abstracted.

Comment (dkarlovi) [reply src/Elasticsearch/Extension/SearchExtension.php]: @sstok So what do you think about this as it is? What changes would you like to see to get this merged?

**Comment by dkarlovi** [comment]
Still required:
1. use limit / offset from query builder, remove them from query builder  
2. require EXACT order from QueryBuilder

Reasoning: As the search engine is in charge of determining what results match the query and what's their significance in the query context, **Elasticsearch is completely in charge** of both deciding **which entities** to display and **in what order**. Doctrine is used only as a storage in this case and should not interfere with the result set, only provide it.

**Comment by sstok** [comment]
Can you reopen this against https://github.com/rollerworks/search 👍

**Pull request closed without merging**